lib/process-services-cloud/src/lib/services/web-socket.service.ts

```diff
@@ -0,0 +1,196 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { createClient } from 'graphql-ws';
+import { Inject, Injectable } from '@angular/core';
+import { AppConfigService, AuthenticationService } from '@alfresco/adf-core';
+import { GraphQLWsLink } from '@apollo/client/link/subscriptions';
+import { WebSocketLink } from '@apollo/client/link/ws';
+import {
+    DefaultContext,
+    FetchResult,
+    from,
+    HttpLink,
+    InMemoryCache,
+    InMemoryCacheConfig,
+    NextLink,
+    Operation,
+    split,
+    SubscriptionOptions
+} from '@apollo/client/core';
+import { getMainDefinition } from '@apollo/client/utilities';
+import { Kind, OperationTypeNode } from 'graphql';
+import { Apollo } from 'apollo-angular';
+import { onError } from '@apollo/client/link/error';
+import { RetryLink } from '@apollo/client/link/retry';
+import { Observable } from 'rxjs/internal/Observable';
+import { switchMap, take, tap } from 'rxjs/operators';
+import { FeaturesServiceToken, IFeaturesService } from '@alfresco/adf-core/feature-flags';
+
+interface serviceOptions {
+    apolloClientName: string;
+    wsUrl: string;
+    httpUrl?: string;
+    subscriptionOptions: SubscriptionOptions;
+}
+
+@Injectable({
+    providedIn: 'root'
+})
+export class WebSocketService {
+    private host = '';
+    private subscriptionProtocol: 'graphql-ws' | 'transport-ws' = 'transport-ws';
+    private wsLink: GraphQLWsLink | WebSocketLink;
+    private httpLink: HttpLink;
+
+    constructor(
+        private readonly apollo: Apollo,
+        private readonly appConfigService: AppConfigService,
+        private readonly authService: AuthenticationService,
+        @Inject(FeaturesServiceToken) private featuresService: IFeaturesService
+    ) {
+        this.host = this.appConfigService.get('bpmHost', '');
+    }
+
+    public getSubscription<T>(options: serviceOptions): Observable<FetchResult<T>> {
+        const { apolloClientName, subscriptionOptions } = options;
+        this.authService.onLogout.pipe(take(1)).subscribe(() => {
+            if (this.apollo.use(apolloClientName)) {
+                this.apollo.removeClient(apolloClientName);
+            }
+        });
+
+        return this.featuresService.isOn$('studio-ws-graphql-subprotocol').pipe(
+            tap((isOn) => {
+                if (isOn) {
+                    this.subscriptionProtocol = 'graphql-ws';
+                }
+            }),
+            switchMap(() => {
+                if (this.apollo.use(apolloClientName) === undefined) {
+                    this.initSubscriptions(options);
+                }
+                return this.apollo.use(apolloClientName).subscribe<T>({ errorPolicy: 'all', ...subscriptionOptions });
```

Comment:
I think it would be better to return the subcriber instead of the Observable so we can directly subscribe to getSubscription()